client/lib/create-cached-selector/index.js

```diff
@@ -0,0 +1,102 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, forEach, some, isFunction } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/*
+ * A map that is Weak with objects but Strong with primitives
```

Comment:
this is a great description

Reply:
😊

Reply:
why do we have this? It strikes me as a bad idea, as anything that ends up in the strong map is effectively a memory leak. The point of using a WeakMap is to weakly hold the keys so that when they go out of scope elsewhere, we can GC them. If we allow some keys to end up in the strong map, they're never be GC'd.

For true primitives, that's probably ok, they don't take up much space, but then there's also not much benefit to caching them this way. For strings, it could get ugly, as they're effectively unbounded in size. 

So why do we need this?

Reply:
I was torn on whether or not to allow primitives. 
There are perfectly valid use-cases for having a primitive as a dependent.
Imagine state looked like this:

```
state = {
  priceOfSite: 5.40,
  sites: [ ... ],
  ...
}
```

If primitives weren't allowed, then you wouldn't be able to write a cached selector that used `priceOfSite`.  For example:

```js
// would be invalid without MixedMap
createCachedSelector( 
  ({ sites, price }) =>return price * sites.length,
  state => ({ sites: state.sites, price: state.priceOfSite }) 
)
```

---

The solutions i could think of are:
1. restrict this cachedSelector function to to selectors that depend on objects within state, no primitives allowed.
2. `MixedMap`
3. selector-ception: Do something clever with primitives. If we notice that a dependent is a primitive, then create a cached function for it so that it will return the value wrapped in an object `{ value: 'primitive-value' }` and make sure to return a referentially equal one whenever the value hasn't changed.

Reply:
@blowery we can't use primitive values as keys in `WeakMap`s sadly. that is, this is getting as much out of the `WeakMap` as we can and only falling back to classical maps when JavaScript won't let us use the `WeakMap`

our alternative is probably to skip caching these values